File: Philosopher's Stone/API/C++/PhilosophersStone.cpp
#include "PhilosophersStone.h"

PhilosophersStone::PhilosophersStone() {

}

PhilosophersStone::~PhilosophersStone() {

}

void PhilosophersStone::tag(std::string tag) {

	if(!isTagged(tag))
		tags.push_back(tag);
}

void PhilosophersStone::removeTag(std::string tag) {

	for(unsigned int i = 0; i < tags.size(); i++) {

		if(tags.at(i).compare(tag) == 0) {

			tags.erase(tags.begin() + i);

			return;
		}
	}
}

std::vector<std::string> PhilosophersStone::getTags() {
	return tags;
}

bool PhilosophersStone::isTagged(std::string tag) {

	for(unsigned int i = 0; i < tags.size(); i++) {

		if(tags.at(i).compare(tag) == 0)
			return true;
	}

	return false;
}

void PhilosophersStone::publiclyConnect(PhilosophersStone* connection) {

	if(isPubliclyConnected(connection))
		return;

	publicConnections.push_back(connection);
}

void PhilosophersStone::privatelyConnect(PhilosophersStone* connection) {

	if(isPrivatelyConnected(connection))
		return;

	privateConnections.push_back(connection);
}

void PhilosophersStone::publiclyConnectMutually(PhilosophersStone* connection) {
	publiclyConnect(connection);
	connection->publiclyConnect(this);
}

void PhilosophersStone::privatelyConnectMutually(PhilosophersStone* connection) {
	privatelyConnect(connection);
	connection->privatelyConnect(this);
}

void PhilosophersStone::disconnect(PhilosophersStone* connection) {

	int index = getIndexOfPublicConnection(connection);

	if(index != -1) {

		publicConnections.erase(publicConnections.begin() + index);

		return;
	}

	index = getIndexOfPrivateConnection(connection);

	if(index != -1) {

		privateConnections.erase(privateConnections.begin() + index);

		return;
	}
}

void PhilosophersStone::disconnectMutually(PhilosophersStone* connection) {
	disconnect(connection);
	connection->disconnect(this);
}

bool PhilosophersStone::isConnected(PhilosophersStone* connection) {
	return isPubliclyConnected(connection) || isPrivatelyConnected(connection);
}

bool PhilosophersStone::isPubliclyConnected(PhilosophersStone* connection) {

	for(unsigned int i = 0; i < publicConnections.size(); i++) {

		if(publicConnections.at(i) == connection)
			return true;
	}

	return false;
}

bool PhilosophersStone::isPrivatelyConnected(PhilosophersStone* connection) {

	for(unsigned int i = 0; i < privateConnections.size(); i++) {

		if(privateConnections.at(i) == connection)
			return true;
	}

	return false;
}

std::vector<PhilosophersStone*> PhilosophersStone::getConnections() {

	std::vector<PhilosophersStone*> connections;

	connections.resize(publicConnections.size() + privateConnections.size());

	for(unsigned int i = 0; i < publicConnections.size(); i++)
		connections.push_back(publicConnections.at(i));

	for(unsigned int i = 0; i < privateConnections.size(); i++)
		connections.push_back(privateConnections.at(i));

	return connections;
}

std::vector<PhilosophersStone*> PhilosophersStone::getPublicConnections() {
	return publicConnections;
}

std::vector<PhilosophersStone*> PhilosophersStone::getPrivateConnections() {
	return privateConnections;
}

std::vector<PhilosophersStone*> PhilosophersStone::get(std::vector<std::string> tags) {

	std::vector<PhilosophersStone*> path;
	std::vector<PhilosophersStone*> stones;

	getTraverse(true, tags, path, stones);

	return stones;
}

std::vector<PhilosophersStone*> PhilosophersStone::getAtlas() {

	std::vector<PhilosophersStone*> path;
	std::vector<PhilosophersStone*> stones;

	getAtlasTraverse(true, path, stones);

	return stones;
}

bool PhilosophersStone::has(std::vector<std::string> tags) {

	std::vector<PhilosophersStone*> path;

	return hasTraverse(true, tags, path);
}

std::vector<void*> PhilosophersStone::call(std::vector<void*> packet) {

	std::vector<PhilosophersStone*> path;
	std::vector<void*> calls;

	callTraverse(true, packet, path, calls);

	return calls;
}

void* PhilosophersStone::onCall(std::vector<void*> packet) {
	return NULL;
}

int PhilosophersStone::getIndexOfPublicConnection(PhilosophersStone* connection) {

	for(unsigned int i = 0; i < publicConnections.size(); i++) {

		if(publicConnections.at(i) == connection)
			return i;
	}

	return -1;
}

int PhilosophersStone::getIndexOfPrivateConnection(PhilosophersStone* connection) {

	for(unsigned int i = 0; i < privateConnections.size(); i++) {

		if(privateConnections.at(i) == connection)
			return i;
	}

	return -1;
}

void PhilosophersStone::callTraverse(
		bool init,
		std::vector<void*> packet,
		std::vector<PhilosophersStone*> path,
		std::vector<void*> calls) {

	if(traversed(path))
		return;

	path.push_back(this);

	void* call = NULL;

	try {
		call = onCall(packet);
	}

	catch(...) {

	}

	if(call != NULL)
		calls.push_back(call);

	if(init) {

		for(unsigned int i = 0; i < privateConnections.size(); i++)
			privateConnections.at(i)->callTraverse(false, packet, path, calls);
	}

	for(unsigned int i = 0; i < publicConnections.size(); i++)
		publicConnections.at(i)->callTraverse(false, packet, path, calls);
}

void PhilosophersStone::getTraverse(
		bool init,
		std::vector<std::string> tags,
		std::vector<PhilosophersStone*> path,
		std::vector<PhilosophersStone*> stones) {

	if(traversed(path))
		return;

	path.push_back(this);

	if(hasAllTags(tags))
		stones.push_back(this);

	if(init) {

		for(unsigned int i = 0; i < privateConnections.size(); i++)
			privateConnections.at(i)->getTraverse(false, tags, path, stones);
	}

	for(unsigned int i = 0; i < publicConnections.size(); i++)
		publicConnections.at(i)->getTraverse(false, tags, path, stones);
}

void PhilosophersStone::getAtlasTraverse(
		bool init,
		std::vector<PhilosophersStone*> path,
		std::vector<PhilosophersStone*> stones) {

	if(traversed(path))
		return;

	path.push_back(this);
	stones.push_back(this);

	if(init) {

		for(unsigned int i = 0; i < privateConnections.size(); i++)
			privateConnections.at(i)->getTraverse(false, tags, path, stones);
	}

	for(unsigned int i = 0; i < publicConnections.size(); i++)
		publicConnections.at(i)->getTraverse(false, tags, path, stones);
}

bool PhilosophersStone::hasTraverse(
		bool init,
		std::vector<std::string> tags,
		std::vector<PhilosophersStone*> path) {

	if(traversed(path))
		return false;

	path.push_back(this);

	if(hasAllTags(tags))
		return true;

	if(init) {

		for(unsigned int i = 0; i < privateConnections.size(); i++) {

			if(privateConnections.at(i)->hasTraverse(false, tags, path))
				return true;
		}
	}

	for(unsigned int i = 0; i < publicConnections.size(); i++) {

		if(publicConnections.at(i)->hasTraverse(false, tags, path))
			return true;
	}

	return false;
}

bool PhilosophersStone::traversed(std::vector<PhilosophersStone*> path) {

	for(unsigned int i = 0; i < path.size(); i++) {

		if(path.at(i) == this)
			return true;
	}

	return false;
}

bool PhilosophersStone::hasAllTags(std::vector<std::string> tags) {

	for(unsigned int i = 0; i < tags.size(); i++) {

		if(!isTagged(tags.at(i)))
			return false;
	}

	return true;
}

File: Philosopher's Stone/API/JavaScript/PhilosophersStone.js
class PhilosophersStone {

	constructor() {
	
		this.connections = [];
		this.privateConnections = [];
		
		this.tags = [];
	}
	
	onCall(packet) {
		return null;
	}
	
	onIncoming(path) {
		return true;
	}
	
	onOutgoing(path) {
		
		if(path.length > 2) {
			
			var stone = path[path.length - 1];
			
			for(var i = 0; i < this.privateConnections.length; i++) {
			
				if(stone === this.privateConnections[i])
					return false;
			}
		}
	
		return true;
	}
}

function getAtlas(stone, all) {
	
	if(Array.isArray(stone)) {
		
		stone = stone.slice(0);
		
		var atlas = [];
		
		for(var i = 0; i < stone.length; i++)
			atlas = atlas.concat(getAtlas(stone[i], all));
		
		for(var i = 0; i < atlas.length; i++) {
		
			for(var j = 0; j < atlas.length; j++) {
				
				if(atlas[i] === atlas[j]) {
				
					atlas.splice(j, 1);
					
					j--;
				}
			}
		}
		
		return atlas;
	}
	
	else {
		
		var atlas = [stone];
		
		for(var i = 0; i < stone.connections.length; i++) {
			
			var projectedPath = atlas.slice(0);
			projectedPath.push(stone.connections[i]);
			
			if(stone.onOutgoing(projectedPath))
				getAtlasTraversal(stone.connections[i], atlas, all);
		}
		
		return atlas;
	}
}

function getAtlasTraversal(stone, atlas, all) {

	for(var i = 0; i < atlas.length; i++) {
		
		if(atlas[i] === stone)
			return;
	}
	
	var incoming = all;
	
	if(!all) {
		
		try {
			incoming = stone.onIncoming(atlas);
		}
		
		catch(error) {
			
		}
	}
	
	if(incoming)
		atlas.push(stone);
	
	for(var i = 0; i < stone.connections.length; i++) {
		
		var projectedPath = atlas.slice(0);
		projectedPath.push(stone.connections[i]);
		
		var outgoing = all;
		
		if(!all) {
			
			try {
				outgoing = stone.onOutgoing(projectedPath);
			}
			
			catch(error) {
				
			}
		}
		
		if(outgoing)
			getAtlasTraversal(stone.connections[i], atlas);
	}
}

function connect(stone, connection, mutual, private) {

	if(mutual) {
		connect(stone, connection, false, private);
		connect(connection, stone, false, private);
	}

	else {
	
		if(Array.isArray(stone)) {
		
			stone = stone.slice(0);
		
			if(Array.isArray(connection)) {
		
				connection = connection.slice(0);
			
				for(var i = 0; i < stone.length; i++) {
				
					for(var j = 0; j < connection.length; j++)
						connect(stone[i], connection[j], mutual, private);
				}
			}
			
			else {
			
				for(var i = 0; i < stone.length; i++)
					connect(stone[i], connection, mutual, private)
			}
		}
	
		else if(Array.isArray(connection)) {
		
			connection = connection.slice(0);
			
			for(var i = 0; i < connection.length; i++)
				connect(stone, connection[i], mutual, private)
		}
		
		else {
		
			if(isConnected(stone, connection, false))
				return;
				
			stone.connections.push(connection);
			
			if(private && !isConnected(stone, connection, true))
				stone.privateConnections.push(connection);
		}
	}
}

function disconnect(stone, connection, mutual) {

	if(mutual) {
		disconnect(stone, connection, false);
		disconnect(connection, stone, false);
	}

	else {
	
		if(Array.isArray(stone)) {
		
			stone = stone.slice(0);
		
			if(Array.isArray(connection)) {
		
				connection = connection.slice(0);
			
				for(var i = 0; i < stone.length; i++) {
				
					for(var j = 0; j < connection.length; j++)
						disconnect(stone[i], connection[j], mutual);
				}
			}
			
			else {
			
				for(var i = 0; i < stone.length; i++)
					disconnect(stone[i], connection, mutual)
			}
		}
	
		else if(Array.isArray(connection)) {
		
			connection = connection.slice(0);
			
			for(var i = 0; i < connection.length; i++)
				disconnect(stone, connection[i], mutual)
		}
		
		else {
			
			for(var i = 0; i < stone.connections.length; i++) {
			
				if(stone.connections[i] === connection) {
				
					stone.connections.splice(i, 1);
					
					i--;
				}
			}
			
			for(var i = 0; i < stone.privateConnections.length; i++) {
			
				if(stone.privateConnections[i] === connection) {
				
					stone.privateConnections.splice(i, 1);
					
					i--;
				}
			}
		}
	}
}

function tag(stone, tags) {
	
	if(Array.isArray(stone)) {
		
		stone = stone.slice(0);
		
		for(var i = 0; i < stone.length; i++)
			tag(stone[i], tags);
	}
	
	else {
	
		if(Array.isArray(tags)) {
		
			tags = tags.slice(0);
		
			for(var i = 0; i < tags.length; i++) {
				
				var tag = formatTag(tags[i]);
				
				if(!stone.tags.includes(tag))
					stone.tags.push(tag);
			}
		}
	
		else
			stone.tags.push(formatTag(tags));
	}
}

function detag(stone, tags) {
	
	if(Array.isArray(stone)) {
		
		stone = stone.slice(0);
		
		for(var i = 0; i < stone.length; i++)
			detag(stone[i], tags);
	}
	
	else {
	
		if(!Array.isArray(tags))
			tags = [tags];
			
		else
			tags = tags.slice(0);

		for(var i = 0; i < stone.tags.length; i++) {
			
			var tag = formatTag(stone.tags[i]);
		
			for(var j = 0; j < tags.length; j++) {
			
				if(tag == formatTag(tags[j])) {
				
					stone.tags.splice(i, 1);
					
					i--;
					break;
				}
			}
		}
	}
}

function isConnected(stone, connection, mutual, private) {

	if(mutual) {
	
		return
			isConnected(stone, connection, false, private) &&
			isConnected(connection, stone, false, private);
	}
	
	else {
	
		if(Array.isArray(stone)) {
		
			stone = stone.slice(0);
		
			if(Array.isArray(connection)) {
		
				connection = connection.slice(0);
				
				for(var i = 0; i < stone.length; i++) {
				
					for(var j = 0; j < connection.length; j++) {
					
						if(!isConnected(stone[i], connection[j], mutual, private))
							return false;
					}
				}
				
				return true;
			}
			
			else {
			
				for(var i = 0; i < stone.length; i++) {
				
					if(!isConnected(stone[i], connection, mutual, private))
						return false;
				}
						
				return true;
			}
		}
	
		else if(Array.isArray(connection)) {
		
			connection = connection.slice(0);
			
			for(var i = 0; i < connection.length; i++) {
			
				if(!isConnected(stone, connection[i], mutual, private))
					return false;
			}
					
			return true;
		}
		
		else {
		
			var connections =
				private ?
					stone.privateConnections :
					stone.connections;
			
			for(var i = 0; i < connections.length; i++) {
				
				if(connections[i] === connection)
					return true;
			}
			
			return false;
		}
	}
}

function isTagged(stone, tags) {
	
	if(Array.isArray(stone)) {
		
		store = store.slice(0);
		
		for(var i = 0; i < stone.length; i++) {
			
			if(!isTagged(stone[i]))
				return false;
		}
		
		return true;
	}
	
	else {
	
		if(!Array.isArray(tags))
			tags = [tags];
	
		for(var i = 0; i < tags.length; i++) {
			
			var found = false;
			
			for(var j = 0; j < stone.tags.length; j++) {
				
				if(formatTag(tags[i]) == formatTag(stone.tags[j])) {
					
					found = true;
					
					break;
				}
			}
			
			if(!found)
				return false;
		}
		
		return true;
	}
}

function get(stones, tags) {
	
	var tagged = [];
	
	for(var i = 0; i < stones.length; i++) {
		
		if(isTagged(stones[i], tags))
			tagged.push(stones[i]);
	}
	
	return tagged;
}

function call(stones, packet) {
	
	var call = [];
	
	for(var i = 0; i < stones.length; i++) {
		
		try {
			
			var value = stones[i].onCall(packet);
			
			if(value != null)
				call.push(value);
		}
		
		catch(error) {
			
		}
	}
	
	return call;
}

function formatTag(tag) {
	
	var newTag = tag.toLowerCase();
	
	for(var i = 0; i < tag.length; i++) {
		
		if(newTag.charAt(i) == " " ||
			newTag.charAt(i) == "\t" ||
			newTag.charAt(i) == "\n") {
		
			newTag = newTag.substring(0, i) + newTag.substring(i + 1);
			
			i--;
		}
	}
	
	return newTag;
}

module.exports = {

	PhilosophersStone,
	getAtlas,
	getAtlasTraversal,
	connect,
	disconnect,
	tag,
	detag,
	isConnected,
	isTagged,
	call,
	get,
	formatTag
};